[Enhancement #521] Allow setting primary language on vocabulary level

lukaskabc opened the pull request:
Resolves #521

- Adds primary language selector to vocabulary edit and creation form
- Vocabulary language is now also pre-selected when creating a new term
- Vocabulary and Terms are prevented from saving when label translation in the vocabulary primary language is missing
- WebSocket is now also monitored for notifications from text analysis failure

lukaskabc committed:
[Enhancement #521] Add option to set primary language for Vocabulary

lukaskabc committed:
[Enhancement #521] Add primary language selection to vocabulary creation form

lukaskabc committed:
[Enhancement #521] Use language native name in vocabulary primary language selection to maintain consistency and preselect vocabulary primary language for term creation

lukaskabc committed:
[Enhancement #521] Delete empty language when primary language is switched during vocabulary creation

lukaskabc committed:
[Enhancement #521] Update language translation selection during vocabulary and term creation/modification, automatically select vocabulary language, prevent removing of primary language translation

lukaskabc committed:
[Enhancement #521] remove unused porperties from TermMetadataEdit, VocabularyEdit and VocabularySummary; update tests with changes to vocabulary primary language

lukaskabc committed:
[Enhancement #521] receive ws message on text analysis failure

lukaskabc committed:
[Enhancement #521] Use strong equality for checking current language

ledsoft commented on src/component/term/TermDetail.tsx:
```diff
@@ -92,6 +92,7 @@ export interface TermDetailState extends EditableComponentState {
 
 export function resolveInitialLanguage(props: CommonTermDetailProps) {
   const { term, configuredLanguage, locale } = props;
+  if (props.vocabulary.primaryLanguage) return props.vocabulary.primaryLanguage;
```
Please use `{}` around if blocks

ledsoft commented on src/component/annotator/TextAnalysisInvocationButton.tsx:
```diff
@@ -63,11 +64,21 @@ export class TextAnalysisInvocationButton extends React.Component<
+      message.headers.destination ===
+      Constants.WEBSOCKET_ENDPOINT.VOCABULARIES_TEXT_ANALYSIS_FAILED
+    ) {
+      // strips double quotes
```
Why?

lukaskabc replied:
I can keep them if you want
quotes stripped:
<img width="485" height="72" alt="{4F50A857-E435-44A2-B80C-9C7A11FE3C84}" src="https://github.com/user-attachments/assets/db627a17-ec0a-4dfb-8d07-48b4c5d69775" />
quotes kept:
<img width="492" height="63" alt="{E7B5CC39-295C-491B-A0DA-DC37EC329399}" src="https://github.com/user-attachments/assets/3e6d8f7c-6648-44f1-b18d-9300b386fe17" />

I could replace the regex with a substring to keep it unified with the stripping above.

ledsoft replied:
Ok, in that case remove them and disregard my comment.

ledsoft commented on src/component/annotator/TextAnalysisInvocationButton.tsx:
Could you rewrite this to a functional component with hooks? We are trying to move away from class components when working on existing code (and the class is not too complex).

ledsoft commented on src/component/term/TermMetadataCreate.tsx:
```diff
@@ -59,12 +64,30 @@ export class TermMetadataCreate extends React.Component<
     this.props.onCreate(t, false);
   };
 
+  private isPrimaryLabelMissing = () => {
```
Can use `Utils.notBlank`

ledsoft commented on src/component/term/TermMetadataCreate.tsx:
```diff
@@ -59,12 +64,30 @@ export class TermMetadataCreate extends React.Component<
+        )
+      );
+      return;
+    }
```
I would stick to the validation mechanism we already have in place - it disables the save button. What we can consider (not necessarily for this PR) is showing the validation results to the user in some meaningful way so that they know why the save button is disabled.

ledsoft commented on src/component/term/TermMetadataEdit.tsx:
Same comments as in TermMetadataCreate apply here.

ledsoft commented on src/component/vocabulary/CreateVocabularyForm.tsx:
```diff
@@ -128,13 +151,61 @@ const CreateVocabularyForm: React.FC<CreateVocabularyFormProps> = ({
+
+  const isPrimaryLabelMissing = () => {
+    return (
+      label[primaryLanguage] == null || label[primaryLanguage].trim() === ""
```
Use `Utils.notBlank`

ledsoft commented on src/component/vocabulary/CreateVocabularyForm.tsx:
```diff
@@ -108,10 +117,24 @@ const CreateVocabularyForm: React.FC<CreateVocabularyFormProps> = ({
+        )
+      );
+      return;
+    }
```
Once again, I would stick to the existing validation mechanism.

ledsoft commented on src/component/vocabulary/VocabularyEdit.tsx:
```diff
@@ -91,7 +121,40 @@ export class VocabularyEdit extends React.Component<
+      );
+      return;
+    }
+
```
Again - existing validation should be used.

ledsoft commented on src/component/vocabulary/CreateVocabularyForm.tsx:
```diff
@@ -194,6 +265,28 @@ const CreateVocabularyForm: React.FC<CreateVocabularyFormProps> = ({
                   />
                 </Col>
               </Row>
+              <Row>
```
Let's move this just below the vocabulary description textarea (applies to vocabulary edit as well).

ledsoft reviewed (commented):
Some inline comments. I also encountered an issue when I created an English vocabulary on Czech instance - it does not display the description of the vocabulary (the description does exist and when I switch to edit and back it is displayed):
<img width="1659" height="497" alt="image" src="https://github.com/user-attachments/assets/52f48129-4db9-42ee-9d5e-29329d23e4f0" />

lukaskabc committed:
[Enhancement #521] Fix vocabulary description not displaying in vocabulary detail with a different language

lukaskabc committed:
[Ref] Rewrite TextAnalysisInvocationButton as functional component

lukaskabc committed:
[Enhancement #521] Unify term and vocabulary validation by disabling the save button when validation is violated

lukaskabc committed:
[Enhancement #521] Add language selector to modal for term creation from file annotator.

lukaskabc committed:
[Enhancement #521] Move vocabulary primary language selector under the description in vocabulary edit

lukaskabc committed:
[Enhancement #521] Update annotator to use file language when marking a term definition

lukaskabc committed:
[Enhancement #521] Update tests to reflect validation changes and rewrite to functional component

lukaskabc committed:
[Ref] Remove unused language translations

lukaskabc commented:
I also noticed that it is possible to create a term from the annotator. I updated it so it respects the file language and added the language selector, because otherwise it would not be possible to create a term while annotating files with a different language than the vocabulary's primary one.

I'll test it further tomorrow

ledsoft committed:
Merge branch 'development' into lukaskabc/Enhancement-ui-521

ledsoft reviewed (approved):
LGTM
